hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/UpgradeSubcommand.java

```diff
@@ -0,0 +1,194 @@
+    }
+
+    final Pair<Integer, Integer> layoutVersion =
+        upgradeChecker.getLayoutVersion(configuration);
+    final Integer softwareLayoutVersion = layoutVersion.getLeft();
+    final Integer metadataLayoutVersion = layoutVersion.getRight();
+    final int needLayoutVersion =
+        HDDSLayoutFeature.DATANODE_SCHEMA_V3.layoutVersion();
+
+    if (metadataLayoutVersion < needLayoutVersion ||
+        softwareLayoutVersion < needLayoutVersion) {
+      out().println(String.format(
+          "Please upgrade your software version, no less than %s," +
+              " current metadata layout version is %d," +
+              " software layout version is %d",
+          HDDSLayoutFeature.DATANODE_SCHEMA_V3.name(),
+          metadataLayoutVersion, softwareLayoutVersion));
+      return null;
+    }
+
+    if (!Strings.isNullOrEmpty(volume)) {
+      configuration.set(ScmConfigKeys.HDDS_DATANODE_DIR_KEY, volume);
+    }
+
+    DatanodeDetails dnDetail =
+        UpgradeUtils.getDatanodeDetails(configuration);
+
+    final HddsProtos.NodeOperationalState opState =
+        dnDetail.getPersistedOpState();
+
+    if (!opState.equals(HddsProtos.NodeOperationalState.IN_MAINTENANCE)) {
+      out().println("This command requires the datanode's " +
+          "NodeOperationalState to be IN_MAINTENANCE, currently is " +
+          opState);
+      return null;
+    }
+
+    List<HddsVolume> allVolume =
+        upgradeChecker.getAllVolume(dnDetail, configuration);
+
+    final List<File> volumeDBPath = upgradeChecker.getVolumeDBPath(allVolume);
+
+    if (volumeDBPath.isEmpty()) {
+      out().println("No volume db store exists, need finalizae data node.");
+      return null;
+    }
+
+    for (HddsVolume hddsVolume : allVolume) {
+      if (UpgradeChecker.checkAlreadyMigrate(hddsVolume)) {
+        out().println("Volume " + hddsVolume.getVolumeRootDir() +
+            " it's already upgraded, skip it.");
+        return null;
+      }
+    }
+
+    if (!yes) {
+      Scanner scanner = new Scanner(new InputStreamReader(
+          System.in, StandardCharsets.UTF_8));
+      System.out.println(
+          "The volume db store, will be automatically backup," +
+              " should you continue to upgrade?");
+      boolean confirm = scanner.next().trim().equals("yes");
+      scanner.close();
+      if (!confirm) {
+        return null;
+      }
+    }
+
+    final Map<File, Exception> backupExceptions =
+        upgradeChecker.dbBackup(volumeDBPath);
+
+    if (!backupExceptions.isEmpty()) {
+      final String backupFailDbPath = Arrays.toString(
+          backupExceptions.keySet().stream().map(File::getAbsolutePath)
+              .distinct().toArray());
+
+      out().println("Db store path: " + backupFailDbPath +
+          " backup fail, please check.");
+      return null;
+    }
+
+    // do upgrade
+    final UpgradeManager upgradeManager = new UpgradeManager();
+    upgradeManager.run(configuration);
+    return null;
```

Comment:
Shall we delete the backup DB files before return?

Reply:
> Shall we delete the backup DB files before return?

I think we can keep it, the db file won't be very big, might as well delete it manually after a while as it can be observed for a while and the data can be rolled back in case of errors